tests/room_keys_test.go

```diff
@@ -91,10 +91,77 @@ func TestRoomKeyIsCycledOnDeviceLogout(t *testing.T) {
 	})
 }
 
+// The room key is cycled when `rotation_period_msgs` is met (default: 100).
+//
+// This test ensures we change the m.room_key when we have sent enough messages,
+// where "enough" means the value set in the `m.room.encryption` event under the
+// `rotation_period_msgs` property.
+//
+// If the key were not changed, someone who stole the key would have access to
+// future messages.
+func TestRoomKeyIsCycledAfterEnoughMessages(t *testing.T) {
+	ClientTypeMatrix(t, func(t *testing.T, clientTypeA, clientTypeB api.ClientType) {
+		// Given a room containing Alice and Bob
+		tc := CreateTestContext(t, clientTypeA, clientTypeB)
+		roomID := tc.CreateNewEncryptedRoom(
+			t,
+			tc.Alice,
+			"trusted_private_chat",
+			[]string{tc.Bob.UserID},
+			map[string]interface{}{
+				"algorithm":            "m.megolm.v1.aes-sha2",
+				"rotation_period_msgs": 5,
+			},
+		)
+		tc.Bob.MustJoinRoom(t, roomID, []string{clientTypeA.HS})
+
+		tc.WithAliceAndBobSyncing(t, func(alice, bob api.Client) {
+			// And some messages were sent, but not enough to trigger resending
+			for i := 0; i < 4; i++ {
+				wantMsgBody := "Before we hit the threshold"
+				waiter := bob.WaitUntilEventInRoom(t, roomID, api.CheckEventHasBody(wantMsgBody))
+				alice.SendMessage(t, roomID, wantMsgBody)
+				waiter.Wait(t, 5*time.Second)
+			}
+
+			// Sniff calls to /sendToDevice to ensure we see the new room key being sent.
+			ch := make(chan deploy.CallbackData, 10)
+			callbackURL, close := sniffToDeviceEvent(t, tc.Deployment, ch)
+			defer close()
+			tc.Deployment.WithMITMOptions(t, map[string]interface{}{
+				"callback": map[string]interface{}{
+					"callback_url": callbackURL,
+					"filter":       "~u .*\\/sendToDevice.*",
+				},
+			}, func() {
+				wantMsgBody := "This one hits the threshold"
+				// When we send two messages (one to hit the threshold and one
+				// to pass it)
```

Comment:
Explain _why_ we do both requests inside the MITM'd section of the test. Critically, it's valid for the client to cycle the key:
 - eagerly as soon as the threshold is reached
 - lazily on the next message that would take the count above the threshold

We should not dictate which one is more correct, as they both are, hence we do 2 requests in the MITM'd section.